### PR TITLE
MOTECH-1956: HL7 task and config integration

### PR DESCRIPTION
**Pull request by tstalka**
- loading hl7 recipients
- building task actions
- loading templates

### CODE REVIEW

**Comment by tstalka** [comment]
test this please

File: ipf/pom.xml [hunk 1]
```diff
@@ -71,7 +90,8 @@
                         <Context-Path>ipf</Context-Path>
                         <Resource-Path>ipf/resources</Resource-Path>
                         <Export-Package>
-                            org.motechproject.ipf.service;version=${project.version}
+                            org.motechproject.ipf.service;version=${project.version},
```
Comment (pgesek) [inline ipf/pom.xml]: domain should be placed before service, alphabetically

File: ipf/src/main/java/org/motechproject/ipf/domain/IPFRecipient.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package org.motechproject.ipf.domain;
+
+public class IPFRecipient {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/domain/IPFRecipient.java:6]: javadoc

File: ipf/src/main/java/org/motechproject/ipf/domain/IPFTemplate.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package org.motechproject.ipf.domain;
+
+import org.apache.commons.lang.ArrayUtils;
+import org.motechproject.mds.annotations.Entity;
+import org.motechproject.mds.annotations.Field;
+
+import java.util.Map;
+
+@Entity
+public class IPFTemplate {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/domain/IPFTemplate.java:13]: javadoc

File: ipf/src/main/java/org/motechproject/ipf/handler/IPFActionEventHandler.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.motechproject.ipf.util.Constants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.Map;
+
+@Service
+public class IPFActionEventHandler {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/handler/IPFActionEventHandler.java:36]: javadoc

File: ipf/src/main/java/org/motechproject/ipf/handler/IPFActionEventHandler.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import org.motechproject.ipf.service.IPFTemplateService;
+import org.motechproject.ipf.util.Constants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.Map;
+
+@Service
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/handler/IPFActionEventHandler.java]: @ Component fits better here

File: ipf/src/main/java/org/motechproject/ipf/service/IPFRecipientsService.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package org.motechproject.ipf.service;
+
+import org.motechproject.ipf.domain.IPFRecipient;
+
+import java.util.Collection;
+
+public interface IPFRecipientsService {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/service/IPFRecipientsService.java:12]: javadoc, javadoc methods

File: ipf/src/main/java/org/motechproject/ipf/service/IPFTemplateDataService.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package org.motechproject.ipf.service;
+
+import org.motechproject.ipf.domain.IPFTemplate;
+import org.motechproject.mds.annotations.Lookup;
+import org.motechproject.mds.annotations.LookupField;
+import org.motechproject.mds.service.MotechDataService;
+
+public interface IPFTemplateDataService  extends MotechDataService<IPFTemplate> {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/service/IPFTemplateDataService.java:11]: javadoc

File: ipf/src/main/java/org/motechproject/ipf/service/IPFTemplateService.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package org.motechproject.ipf.service;
+
+import org.motechproject.ipf.domain.IPFTemplate;
+
+import java.util.List;
+
+public interface IPFTemplateService {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/service/IPFTemplateService.java]: javadoc, javadoc methods

File: ipf/src/main/java/org/motechproject/ipf/service/TaskService.java [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package org.motechproject.ipf.service;
+
+public interface TaskService {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/service/TaskService.java]: javadoc, javadoc methods

File: ipf/src/main/java/org/motechproject/ipf/service/impl/IPFRecipientsServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+import javax.annotation.PostConstruct;
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Service("ipfRecipientsService")
+public class IPFRecipientsServiceImpl implements IPFRecipientsService {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/service/impl/IPFRecipientsServiceImpl.java:29]: javadoc

File: ipf/src/main/java/org/motechproject/ipf/service/impl/IPFTemplateServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.regex.Pattern;
+
+@Service("ipfTemplateService")
+public class IPFTemplateServiceImpl implements IPFTemplateService {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/service/impl/IPFTemplateServiceImpl.java]: javadoc

File: ipf/src/main/java/org/motechproject/ipf/service/impl/IPFTemplateServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    @Autowired
+    @Qualifier("ipfSettings")
+    private SettingsFacade settingsFacade;
+
+    @Autowired
+    private IPFTemplateDataService ipfTemplateDataService;
+
+    @PostConstruct
+    public void init() {
+        Properties properties = settingsFacade.getProperties(Constants.IPF_PROPERTIES_FILE);
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/service/impl/IPFTemplateServiceImpl.java]: What are we exactly doing here? Isn't just storing templates in the database enough? Why are we reading files from the fs?

File: ipf/src/main/java/org/motechproject/ipf/service/impl/TaskServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.PostConstruct;
+import java.util.Collection;
+import java.util.List;
+
+@Service
+public class TaskServiceImpl implements TaskService {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/service/impl/TaskServiceImpl.java]: javadoc

File: ipf/src/main/java/org/motechproject/ipf/task/IpfChannelRequestBuilder.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+import org.motechproject.tasks.contract.ChannelRequest;
+import org.osgi.framework.BundleContext;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class IpfChannelRequestBuilder {
```
Comment (pgesek) [inline ipf/src/main/java/org/motechproject/ipf/task/IpfChannelRequestBuilder.java:25]: javadoc

File: ipf/src/main/resources/ipf.properties [hunk 15]
```diff
@@ -0,0 +1 @@
+ipf.templates=/home/motech/templates
```
Comment (pgesek) [inline ipf/src/main/resources/ipf.properties]: can we remove this now? Along with the file handling code?

File: ipf/src/test/java/org/motechproject/ipf/service/impl/IPFTemplateServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Test
+    public void shouldLoadTemplates() throws IOException {
+        File[] files = generateFiles();
+
+        ipfTemplateService.init();
+        deleteFiles(files);
+
+        Mockito.verify(ipfTemplateDataService, times(2)).createOrUpdate(any(IPFTemplate.class));
```
Comment (pgesek) [inline ipf/src/test/java/org/motechproject/ipf/service/impl/IPFTemplateServiceTest.java]: just static import Mockit methods

File: ipf/src/test/java/org/motechproject/ipf/service/impl/IPFTemplateServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Test
+    public void shouldLoadTemplateData() throws IOException {
+        File tempFile = new File(TEMPLATES_DIR + File.separator + "IPF_motech_template_1.xml");
+        File propsFile = new File(TEMPLATES_DIR + File.separator + "IPF_motech_template_1.properties");
+        tempFile.getParentFile().mkdirs();
+        tempFile.createNewFile();
+        propsFile.createNewFile();
+
+        byte[] templateData = IOUtils.toByteArray(getClass().getResourceAsStream("/IPF_test_template.xml"));
```
Comment (pgesek) [inline ipf/src/test/java/org/motechproject/ipf/service/impl/IPFTemplateServiceTest.java]: close streams

File: ipf/src/test/java/org/motechproject/ipf/service/impl/IPFTemplateServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+        IOUtils.write(templateData, new FileOutputStream(tempFile));
+        byte[] propertiesData = IOUtils.toByteArray(getClass().getResourceAsStream("/IPF_test_template.properties"));
```
Comment (pgesek) [inline ipf/src/test/java/org/motechproject/ipf/service/impl/IPFTemplateServiceTest.java]: close streams

**Pull request merged**